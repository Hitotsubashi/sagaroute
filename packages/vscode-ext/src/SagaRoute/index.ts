import BundledSagaRoute, { RoutingOption } from '@sagaroute/react';
import * as vscode from 'vscode';
import mergeOption from '@sagaroute/react/lib/utils/mergeOption';
import loggingHooks from './hooks/logging';
import writingHooks from './hooks/writing';
import cloneDeep from 'lodash/cloneDeep';
import statisticPathHooks from './hooks/statistic-path';
import getWarningManager from '../WarningManager';
import getLogging from '../Logging';

const workspaceRootFolderPath = vscode.workspace.workspaceFolders![0].uri.fsPath;

const mergedHooks = [loggingHooks, writingHooks, statisticPathHooks].reduce(
  (pre, cur) => mergeOption(pre, { hooks: cur }),
  {} as RoutingOption,
).hooks;

const sagaRouteConfig: RoutingOption = {
  rootPath: workspaceRootFolderPath,
  hooks: mergedHooks,
  onWarning(message) {
    const warningManager = getWarningManager();
    warningManager.addMessage(message);
  },
};

function getSagaRouteConfig() {
  return cloneDeep(sagaRouteConfig);
}

class Sagaroute extends BundledSagaRoute {
  error?: boolean;
}

let sagaRoute: Sagaroute;
// const dependencyName = '@sagaroute/react';

function getSagarouteClass(): typeof Sagaroute {
  return BundledSagaRoute;
  // const logging = getLogging();
  // try {
  //   if (!vscode.workspace.isTrusted) {
  //     logging.logMessage('This workspace is not trusted. Using the bundled sagaroute.');
  //     return BundledSagaRoute;
  //   }
  //   const stats = fs.statSync(path.join(workspaceRootFolderPath, 'package.json'));
  //   if (stats?.isFile()) {
  //     const packageJSON = require(path.join(workspaceRootFolderPath, 'package.json'));
  //     const version =
  //       packageJSON.devDependencies?.[dependencyName] || packageJSON.dependencies?.[dependencyName];
  //     if (version) {
  //       const localSagaroutePath = path.join(
  //         workspaceRootFolderPath,
  //         'node_modules',
  //         dependencyName,
  //       );
  //       if (fs.statSync(localSagaroutePath)) {
  //         logging.logMessage(`Using the local sagaroute:${version}.`);
  //         return require(localSagaroutePath).default;
  //       } else {
  //         logging.logMessage(
  //           `The file [package.json] has "${dependencyName}" in dependency but not found in node_modules. Using the bundled sagaroute.`,
  //         );
  //         return BundledSagaRoute;
  //       }
  //     } else {
  //       logging.logMessage(
  //         `The file [package.json] doesn't has "${dependencyName}" in either dependency or devDependency. Using the bundled sagaroute.`,
  //       );
  //       return BundledSagaRoute;
  //     }
  //   } else {
  //     logging.logMessage('The file [package.json] is not exist. Using the bundled sagaroute.');
  //     return BundledSagaRoute;
  //   }
  // } catch (err) {
  //   logging.logMessage(`ERROR: ${err}. Using the bundled sagaroute.`, 'ERROR');
  //   return BundledSagaRoute;
  // }
}

export default function getSagaRoute() {
  const logging = getLogging();
  if (!sagaRoute) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const SagaRoute = getSagarouteClass();
    try {
      sagaRoute = new SagaRoute(getSagaRouteConfig());
    } catch (error: any) {
      logging.logMessage(`Init Sagaroute Error: ${error.message}.`, 'ERROR');
      vscode.window.showErrorMessage(error.message);
      sagaRoute.error = true;
    }
  }
  return sagaRoute;
}

export function rebuildSagaroute() {
  const logging = getLogging();
  try {
    sagaRoute.generateOptions(getSagaRouteConfig());
    sagaRoute.error = false;
  } catch (error: any) {
    logging.logMessage(`Init Sagaroute Error: ${error.message}.`, 'ERROR');
    vscode.window.showErrorMessage(error.message);
    sagaRoute.error = true;
  }
  return sagaRoute;
}
