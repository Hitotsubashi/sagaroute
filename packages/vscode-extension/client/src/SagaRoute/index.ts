import BundledSagaRoute from '@sagaroute/react';
import SagaRoute, { RoutingOption } from '@sagaroute/react';
import * as vscode from 'vscode';
import mergeOption from '@sagaroute/react/lib/utils/mergeOption';
import loggingHooks from './hooks/logging';
import writingHooks from './hooks/writing';
import cloneDeep from 'lodash/cloneDeep';
import statisticPathHooks from './hooks/statistic-path';
import getWarningManager from '../WarningManager';
import getLogging from '../Logging';
import fs from 'fs';
import path from 'path';

const workspaceRootFolderPath = vscode.workspace.workspaceFolders![0].uri.fsPath;

const mergedHooks = [loggingHooks, writingHooks, statisticPathHooks].reduce(
  (pre, cur) => mergeOption(pre, { hooks: cur }),
  {} as RoutingOption,
).hooks;

const sagaRouteConfig: RoutingOption = {
  rootPath: workspaceRootFolderPath,
  hooks: mergedHooks,
  onWarning(message) {
    const warningManager = getWarningManager();
    warningManager.addMessage(message);
  },
};

export function getSagaRouteConfig() {
  return cloneDeep(sagaRouteConfig);
}

let sagaRoute: SagaRoute;
const dependencyName = '@sagaroute/react';

function getSagarouteClass(): typeof BundledSagaRoute {
  const logging = getLogging();
  try {
    if (!vscode.workspace.isTrusted) {
      logging.logMessage('This workspace is not trusted. Using the bundled sagaroute.');
      return BundledSagaRoute;
    }
    const stats = fs.statSync(path.join(workspaceRootFolderPath, 'package.json'));
    if (stats?.isFile()) {
      const packageJSON = require(path.join(workspaceRootFolderPath, 'package.json'));
      const version =
        packageJSON.devDependencies?.[dependencyName] || packageJSON.dependencies?.[dependencyName];
      if (version) {
        const localSagaroutePath = path.join(
          workspaceRootFolderPath,
          'node_modules',
          dependencyName,
        );
        if (fs.statSync(localSagaroutePath)) {
          logging.logMessage(`Using the local sagaroute:${version}.`);
          return require(localSagaroutePath).default;
        } else {
          logging.logMessage(
            `The file [package.json] has "${dependencyName}" in dependency but not found in node_modules. Using the bundled sagaroute.`,
          );
          return BundledSagaRoute;
        }
      } else {
        logging.logMessage(
          `The file [package.json] doesn't has "${dependencyName}" in either dependency or devDependency. Using the bundled sagaroute.`,
        );
        return BundledSagaRoute;
      }
    } else {
      logging.logMessage('The file [package.json] is not exist. Using the bundled sagaroute.');
      return BundledSagaRoute;
    }
  } catch (err) {
    logging.logMessage(`ERROR: ${err}. Using the bundled sagaroute.`, 'ERROR');
    return BundledSagaRoute;
  }
}

export default function getSagaRoute() {
  if (!sagaRoute) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const SagaRoute = getSagarouteClass();
    sagaRoute = new SagaRoute(getSagaRouteConfig());
  }
  return sagaRoute;
}
