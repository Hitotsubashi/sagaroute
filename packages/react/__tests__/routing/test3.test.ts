import path from 'path';
import SagaRoute from '@/index';
import fs from 'fs';

test('test routing that has same import between route and pagefile.', () => {
  const sagaRoute = new SagaRoute({
    rootPath: path.join(__dirname, 'test-fixtures', 'test3'),
  });
  sagaRoute.routing();
  const context = fs.readFileSync(
    path.join(__dirname, 'test-fixtures', 'test3', 'src', 'routes.tsx'),
    'utf-8',
  );
  expect(context).toMatchSnapshot();
});
