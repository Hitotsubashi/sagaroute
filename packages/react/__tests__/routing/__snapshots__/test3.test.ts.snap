// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test routing that has same import between route and pagefile. 1`] = `
"// @ts-nocheck
// sagaroute-inject:imports
/* injected by sagaroute: start */
import LayoutsIndex from "./layouts/index.tsx";
import Pages404 from "./pages/404.tsx";
import PagesIndex from "./pages/index.tsx";
import PagesUser from "./pages/user.tsx";
import PagesNormal from "./pages/normal.tsx";
/* injected by sagaroute: end */

import React from 'react';

// sagaroute-inject: routes
const routes = [
  {
    "path": "/",
    "element": <LayoutsIndex/>,
    "children": [
      {
        "path": "404",
        "element": <Pages404/>
      },
      {
        "index": true,
        "errorElement": <React.Fragement>
      <div>error...</div>
      <button>back</button>
    </React.Fragement>,
        "element": <PagesIndex/>
      },
      {
        "path": "user1",
        "element": <PagesUser/>
      }
    ]
  },
  {
    "path": "*",
    "element": <PagesNormal/>
  }
];
export default routes;"
`;
